File: lua/problem_29.lua
local function rleEncode(s)
    if #s == 0 then return '' end
    local result = {}
    local count = 1
    for i = 2, #s do
        if s:sub(i, i) == s:sub(i - 1, i - 1) then
            count = count + 1
        else
            result[#result + 1] = tostring(count) .. s:sub(i - 1, i - 1)
            count = 1
        end
    end
    result[#result + 1] = tostring(count) .. s:sub(#s, #s)
    return table.concat(result)
end

local function rleDecode(s)
    local result = {}
    local i = 1
    while i <= #s do
        local j = i
        while j <= #s and s:sub(j, j):match('%d') do j = j + 1 end
        local count = tonumber(s:sub(i, j - 1))
        result[#result + 1] = string.rep(s:sub(j, j), count)
        i = j + 1
    end
    return table.concat(result)
end

local function problem29()
    print("Problem 29:")
    assert(rleEncode("AAAABBBCCDAA") == "4A3B2C1D2A",   "assertion failed")
    assert(rleDecode("4A3B2C1D2A")  == "AAAABBBCCDAA",  "assertion failed")
    assert(rleEncode("A")           == "1A",             "assertion failed")
    assert(rleDecode("1A")          == "A",              "assertion failed")
    assert(rleEncode("")            == "",               "assertion failed")
    print("All Tests Passed")
end

return { problem29 = problem29 }

File: python/problem_29.py
def rle_encode(s: str) -> str:
    if not s:
        return ''
    result = []
    count = 1
    for i in range(1, len(s)):
        if s[i] == s[i - 1]:
            count += 1
        else:
            result.append(f'{count}{s[i-1]}')
            count = 1
    result.append(f'{count}{s[-1]}')
    return ''.join(result)


def rle_decode(s: str) -> str:
    result = []
    i = 0
    while i < len(s):
        j = i
        while j < len(s) and s[j].isdigit():
            j += 1
        count = int(s[i:j])
        result.append(s[j] * count)
        i = j + 1
    return ''.join(result)


def main():
    print("Problem 29:")
    assert rle_encode("AAAABBBCCDAA") == "4A3B2C1D2A"
    assert rle_decode("4A3B2C1D2A")  == "AAAABBBCCDAA"
    assert rle_encode("A")           == "1A"
    assert rle_decode("1A")          == "A"
    assert rle_encode("")            == ""
    print("All Tests Passed")
